File: backend/app/schemas/catalogue.py
from typing import Any

from pydantic import BaseModel, Field


class CatalogueEpisode(BaseModel):
    content_group: str
    title: str
    episode_number: int
    description: str | None = None
    duration_seconds: int | None = None
    video_url: str | None = None
    languages: list[str] = Field(default_factory=list)


class CatalogueSeason(BaseModel):
    season_number: int
    title: str | None = None
    description: str | None = None
    episodes: list[CatalogueEpisode] = Field(default_factory=list)


class CatalogueShow(BaseModel):
    id: int
    title: str
    slug: str
    description: str | None = None
    section: str
    categories: list[str] = Field(default_factory=list)
    release_year: int | None = None
    artworks: dict[str, str | None] = Field(default_factory=dict)
    seasons: list[CatalogueSeason] = Field(default_factory=list)


class CatalogueSection(BaseModel):
    section: str
    shows: list[CatalogueShow] = Field(default_factory=list)


class CatalogueDocument(BaseModel):
    generated_at: str
    version: int
    sections: list[CatalogueSection] = Field(default_factory=list)


class PublishResponse(BaseModel):
    valid: bool
    status: str
    message: str
    catalogue_path: str | None = None
    version: int | None = None
    show_count: int = 0
    episode_count: int = 0
    error_count: int = 0
    errors: list[dict[str, Any]] = Field(default_factory=list)
    warnings: list[dict[str, Any]] = Field(default_factory=list)

File: frontend/src/api/catalogue.ts
import { api } from './client'

export type CatalogueSectionName = 'featured' | 'series' | 'minisodes' | 'songs'

export interface CatalogueEpisode {
  content_group: string
  title: string
  episode_number: number
  description: string | null
  duration_seconds: number | null
  video_url: string | null
  languages: string[] | string
}

export interface CatalogueSeason {
  season_number: number
  title: string | null
  description: string | null
  episodes: CatalogueEpisode[]
}

export interface CatalogueShow {
  id: number
  title: string
  slug: string
  description: string | null
  section: CatalogueSectionName
  categories: string[]
  release_year: number | null
  artworks: Record<string, string | null>
  seasons: CatalogueSeason[]
}

export interface CatalogueSection {
  section: CatalogueSectionName
  shows: CatalogueShow[]
}

export interface CatalogueDocument {
  generated_at: string
  version: number
  sections: CatalogueSection[]
}

export interface CatalogueSearchItem {
  show_id: number
  show_title: string
  show_slug: string
  description: string | null
  section: CatalogueSectionName
  categories: string[]
  release_year: number | null
  season_number: number
  season_title: string | null
  content_group: string
  episode_title: string
  episode_number: number
  episode_description: string | null
  duration_seconds: number | null
  video_url: string | null
  languages: string[]
  artworks: Record<string, string | null>
}

export interface CatalogueSearchResponse {
  items: CatalogueSearchItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
  query: string | null
  category: string | null
  language: string | null
  section: CatalogueSectionName | null
}

export interface CatalogueSearchParams {
  q?: string
  category?: string
  language?: string
  section?: CatalogueSectionName
  page?: number
  page_size?: number
}

export async function getCatalogue(): Promise<CatalogueDocument> {
  const response = await api.get<CatalogueDocument>('/catalog')
  return response.data
}

export async function searchCatalogue(
  params: CatalogueSearchParams,
): Promise<CatalogueSearchResponse> {
  const response = await api.get<CatalogueSearchResponse>('/catalog/search', {
    params: {
      q: params.q || undefined,
      category: params.category || undefined,
      language: params.language || undefined,
      section: params.section || undefined,
      page: params.page ?? 1,
      page_size: params.page_size ?? 12,
    },
  })
  return response.data
}

export function getArtworkUrl(path: string | null | undefined): string | null {
  if (!path) return null
  if (path.startsWith('http://') || path.startsWith('https://')) return path
  return path
}

export function getEpisodeCount(show: CatalogueShow): number {
  return (show.seasons ?? []).reduce(
    (total, season) => total + (season.episodes ?? []).length,
    0,
  )
}

export function getEpisodeLanguages(episode: CatalogueEpisode): string[] {
  return Array.isArray(episode.languages) ? episode.languages : [episode.languages]
}

export function formatDuration(seconds: number | null): string {
  if (!seconds) return 'Runtime unavailable'
  const minutes = Math.floor(seconds / 60)
  const remainingSeconds = seconds % 60
  return `${minutes}m ${String(remainingSeconds).padStart(2, '0')}s`
}
